### PR TITLE
Heap facet

### PR DESCRIPTION
**Pull request by rachitnigam**
Implement apply and get for heapFacets.

### CODE REVIEW

File: js/src/main/scala/HeapValues.scala [hunk 1]
```diff
@@ -218,18 +218,40 @@ trait HeapValues extends slogging.LazyLogging {
     }
   }
 
-  case class HeapFacet(k: Label, v1: HeapValue, v2: HeapValue)
-    extends HeapValue {
-    def get(addr: VHeapRef, field: Value, k: Cont): Config =
-      throw new NotImplementedError(s"Tried to call get on $this")
+  case class HeapFacet(k: Label, v1: HeapValue, v2: HeapValue) extends HeapValue {
+    def mapFacet(v: HeapValue, pc: PC)(f: HeapValue => Expr): Expr = v match {
```
Comment (arjunguha) [inline js/src/main/scala/HeapValues.scala]: Can this be made private?

File: js/src/main/scala/HeapValues.scala [hunk 2]
```diff
@@ -218,18 +218,40 @@ trait HeapValues extends slogging.LazyLogging {
...
+          }
+        }
+      }
+      case _ => f(v)
+    }
+
+    def get(addr: VHeapRef, field: Value, k: Cont): Config = {
+      EConfig(mapFacet(this, store.pc)( (obj0: HeapValue) =>
+          Syntax.mapFacet(field, store.pc)(key0 =>
+              Get(ValueExpr(store.alloc(obj0)), ValueExpr(key0)))), Map.empty, k)
```
Comment (arjunguha) [inline js/src/main/scala/HeapValues.scala]: Something is almost certainly wrong here.  A get operation shouldn't update the store.

File: js/src/main/scala/HeapValues.scala [hunk 3]
```diff
@@ -218,18 +218,40 @@ trait HeapValues extends slogging.LazyLogging {
...
       throw new NotImplementedError(s"Tried to call set on $this")
     def hasOwnProperty(field: Value): Value =
       throw new NotImplementedError(s"Tried to call hasOwnProperty on $this")
-    def apply(actuals: List[Value], k : Cont): Config =
-      throw new NotImplementedError(s"Tried to call apply on $this")
     def invokeNew(self: VHeapRef, args: List[Value], k: Cont): Config =
       throw new NotImplementedError(s"Tried to call new on $this")
+    def apply(actuals: List[Value], k : Cont): Config = {
+      val e = mapFacet(this, store.pc)({ case (f0: HeapValue) =>
+        App(ValueExpr(store.alloc(f0)), actuals.map(arg => ValueExpr(arg)))
```
Comment (arjunguha) [inline js/src/main/scala/HeapValues.scala]: Again, why the allocations? This almost certainly wrong.

**Comment by arjunguha** [comment]
Hold off on merging. Read operations should not write to the store. I think I see why they are necessary, but that probably means that something is wrong in the design of HeapFacet. Think about what re-design would make the writes-in-reads unnecessary.

**Comment by arjunguha** [comment]
Are heap-facets needed to get the Elm demo working?

**Comment by rachitnigam** [comment]
Yes

**Comment by rachitnigam** [comment]
The reason I was having to call `store.alloc` was because the ValueExpr needs to hold a Value which is supposed to be the Facet [(from here)](https://github.com/plasma-umass/Fission/blob/master/js/src/main/scala/JavaScriptCEKLike.scala#L289). Since store.alloc was the only way of creating the VHeapRef, I was using it. 
The correct solving would be storing the VHeapRef in a Map with the HeapValues and get that when needed inside the function.

**Comment by arjunguha** [comment]
Just to confirm, we are closing this because the simple-get branch has an approach that is likely better. Correct?

**Comment by rachitnigam** [comment]
Right. Once we figure a way to correctly implement get, fix the way set works with set and apply on that branch, we can PR it.

**Pull request closed without merging**